### ISSUE TITLE
Unable to undeploy app after deployment with failing healthcheck

### ISSUE THREAD

**Issue by AbaevTM** [spring-cloud/spring-cloud-skipper#964]
Hello, I'm running SCDF on k8s with helm and found issue looking to me as a bug.

- k8s version - 1.16 (AWS EKS eks.1)
- helm chart - https://hub.helm.sh/charts/stable/spring-cloud-data-flow/2.7.1

Steps to reproduce:

1. Create simplest stream with time and log : "time | log"
2. Change _liveness-probe-path_ and _readiness-probe-path_ to value that intentionally breaks healthcheck for time app. 
3. Update stream.
4. Initial version of time app is destroyed and new one is started. After couple of minutes app's state is changed to 'failed'
 and stream's status to 'partial' _**(you can try to perform this step couple of times to generate more failing app versions)**_.
5*. If at this stage you try to undeploy the stream or destroy it both deployments for time and log apps are deleted - this is expected _**(don't try to remove stream when trying to reproduce the issue, I mentioned this step only to show that deletion of the stream at this stage works and SCDF can delete failing app)**_.
5. **For initial stream existing since step 1** change values for _liveness-probe-path_ and _readiness-probe-path_ back for time app to normal. New version of time is started.
6. Previous version is still trying to start, running into crashloopbackoff. 
7. Now destroy, or undeploy the stream. Deployments of healthy apps are deleted, but failing ones are preserved. Also tried `stream all destroy --force` - same results.

If multiple updates are made and app cannot start, all these versions are preserved after stream is destroyed, also occupying resources of k8s cluster.
Only manual deletion with `kubectl  delete deployment appDeploymentName` helps.

In skipper logs I can also observe such lines:

> 2020-07-13 13:40:57.974  INFO 1 --- [eTaskExecutor-2] o.s.c.s.s.d.s.HandleHealthCheckStep      : Release testTimeLogStream-v2 has been DEPLOYED
2020-07-13 13:40:57.974  INFO 1 --- [eTaskExecutor-2] o.s.c.s.s.d.s.HandleHealthCheckStep      : Apps in release testTimeLogStream-v2 are healthy.
2020-07-13 13:40:57.984  INFO 1 --- [eTaskExecutor-2] o.s.c.s.s.d.s.HandleHealthCheckStep      : Deleting changed applications from existing release testTimeLogStream-v1
2020-07-13 13:40:57.995  WARN 1 --- [eTaskExecutor-2] o.s.c.s.s.d.strategies.DeleteStep        : For Release name testTimeLogStream, did not undeploy existing app time as its status is not 'deployed'.

So it looks like it's expected situation with current DeleteStep logic. However to me it looks like bug, as there are not cleaned resources left in k8s but from SCDF perspective stream's health gets back to healthy.

Let me know if I need to provide more details. But setup is pretty basic and the issue can be reproduced easily.

**Comment by jvalkeal** [comment]
I tried this manually with minikube with both scdf 2.5.1 and 2.5.3. Can't reproduce this behaviour(undeploy clears things out) so I'm just wondering if it's something to do with a platform itself or particular k8s version. Can't really see helm chart having any difference per manual kubectl deploy.

cc @chrisjs You have any ideas?

**Comment by AbaevTM** [comment]
@jvalkeal , when you update the app with correct value and new version is deployed do you still see old one? I mean before undeploying stream. Deploy, update (with failing config)... update (with correct config).... check if previous version removed and undeploy after.

**Comment by chrisjs** [comment]
@AbaevTM i tried reproducing based on your steps above and couldn't. i logged what i did below based on what it sounded like you were doing. maybe you can take a look and see if there's something i'm not doing to reproduce properly?

env:
--
* minikube, k8s 1.17.2
* helm 2.7.1 chart
* springcloudstream/time-source-rabbit:2.1.2.RELEASE

issue step 1:
--
`stream create --name tt --definition "time | log" --deploy`

```
tt-log-v1-74d4d7fc4d-l9jzm                     1/1     Running   0          2m25s
tt-time-v1-779fcf696c-gtk9r                    1/1     Running   0          2m25s
```

issue step 2 & 3:
--
```
stream update --name tt --properties "deployer.time.kubernetes.livenessProbePath=/lpp,deployer.time.kubernetes.readinessProbePath=/rpp"
```

```
tt-log-v1-74d4d7fc4d-l9jzm                     1/1     Running   0          3m40s
tt-time-v2-78fdbdc59b-n57lr                    0/1     Running   0          46s
```

* v1 log remains, v1 time terminates, v2 time is created

issue step 4:
--
* let tt-time-v2 container restart 3 times, failing liveness probe, the following appears in shell as the `stream list` status:

```
In the case of multiple apps, some have successfully deployed, while others have not
```

and in `stream info` status:

```
partial
```

issue step "5*":
--
`stream destroy --name tt`

* stream related pods, services, deployments, etc are deleted.

issue step 5:
--
```
stream create --name tt --definition "time | log"
stream deploy --name tt --properties "deployer.time.kubernetes.livenessProbePath=/actuator/health,deployer.time.kubernetes.readinessProbePath=/actuator/info"
```

```
tt-log-v3-6c857c7594-67kmh                     1/1     Running   0          2m24s
tt-time-v3-7d46b97c65-v7j75                    1/1     Running   0          2m24s
```

* no errors were observed in the events table for the time pod

issue step 6:
--
* there are no previous versions as the stream was deleted in issue step "5*"

issue step 7:
--
`stream all destroy`

* stream related pods, services, deployments, etc are deleted.

**Comment by AbaevTM** [comment]
@chrisjs thanks for your check!

Sorry if I wasn't clear enough with description.

In terms of how you described and enumerated steps, where I have issues is after step **5**. I tried to explain that after step 5 you have 2 possible scenarios - **5*** try to delete stream - it works as expected - confirmed and second variation is not to delete stream after 5th step but to continue to update is with other paths to healthcheck. You can do multiple stream updates with incorrect values, too.

And eventually when you change values to normal ones I can observe all previous versions left in k8s.

**Comment by AbaevTM** [comment]
updated description

So steps to reproduce should look like:

>              _ 5* ---> success
>             /
> 1-2-3-4-5-6-7---> failure

**Comment by chrisjs** [comment]
ok i believe i am now able to reproduce what you are seeing:

* do steps 1-4, observe:

```
pod/tt-log-v1-6688fd46db-nm4wm                     1/1     Running            0          48m
pod/tt-time-v2-5c5d78f666-9p7j5                    0/1     CrashLoopBackOff   9          46m
```

* rather than step 5 as listed in my comment above, update again with wrong path:

```
stream update --name tt --properties "deployer.time.kubernetes.livenessProbePath=/lpp2,deployer.time.kubernetes.readinessProbePath=/rpp2"
```

and observe:

```
pod/tt-log-v1-6688fd46db-nm4wm                     1/1     Running            0          50m
pod/tt-time-v2-5c5d78f666-9p7j5                    0/1     CrashLoopBackOff   9          48m
pod/tt-time-v3-94778c9df-rfk5t                     0/1     Running            9          42m
```

* now update with the correct paths:

```
stream update --name tt --properties "deployer.time.kubernetes.livenessProbePath=/actuator/health,deployer.time.kubernetes.readinessProbePath=/actuator/info"
```

and observe:

```
pod/tt-log-v1-6688fd46db-nm4wm                     1/1     Running            0          51m
pod/tt-time-v2-5c5d78f666-9p7j5                    0/1     CrashLoopBackOff   9          49m
pod/tt-time-v3-94778c9df-rfk5t                     0/1     Running            9          43m
pod/tt-time-v4-8676cc6968-84xfc                    1/1     Running            0          12m
```

at this point v2 and v3 are not working but hang around, v3 is working and can be verified via logs.

* try to delete the streams:

```
stream all destroy --force
```

and observe:

```
pod/tt-time-v2-5c5d78f666-9p7j5                    0/1     Running            10         51m
pod/tt-time-v3-94778c9df-rfk5t                     0/1     CrashLoopBackOff   9          45m
```

thanks for clarifying the steps to reproduce

**Comment by AbaevTM** [comment]
@chrisjs thanks for quick response

**Comment by jvalkeal** [comment]
Closing this in favour of #966

**Comment by jvalkeal** [comment]
Better have these as a separate issues vs. #966

**Comment by AbaevTM** [comment]
Hey, @jvalkeal ! Do I understand correctly, that the fix most likely will be delivered in 2.7.0.M1? As #972 and #975 mention it.
